README.md
#npm-lazy-mirror
---

A lazy mirroring local npm server.

[![Build Status](https://travis-ci.org/ActiveState/npm-lazy-mirror.svg?branch=master)](https://travis-ci.org/ActiveState/npm-lazy-mirror)

## About

This package provides a lazy mirroring option for those that:

  * Don't want to mirror the entire couchDB for npmjs.org
  * Don't want to setup cron jobs for those tasks
  * Want to use an in-memory caching server (allocation configurable)
  * Want to easily cache dist tarballs and package metadata to disk

## Install

  * `npm install -g npm-lazy-mirror`

## Run

With CLI flags:

  * `npm-lazy-mirror -p <port> -a <remote_address> -b <bind_address> --cache_dir /npm-data`

With a JSON configuration:

  * `npm-lazy-mirror -C /path/to/config,json`

  See `example/server-config.json` for usage.

**Note:** Your `remote_address` configuration is important, as it is the address used when re-writing
tarball URLs in the metadata. It's certainly always best to use a DNS entry here,
rather than an IP.

## Client configuration

Simply point your local npm config to the lazy mirror (permanent):

    npm config set registry http://localhost:2000/

or per install:

    npm i --registry http://localhost:2000 supertest

## Help

Run `npm-lazy-mirror -h` to see a full list of options.

## Features

  * Caches all tarball / JSON metadata to disk
  * Mirror serves files (200MB max by default) from memory, with a configurable LRU cache.
  * Ability to blacklist packages by semantic versioning specification
  * Option to serve stale resources while the upstream registry is offline
  * Upstream resources are fetched on the fly from the remote registry, the fetching, storing and serving to the client all happen in the same request.
  * Configurable with custom npm registries.
  * HTTP/S proxy support
  * It's Fast and stands up under load. Expect 5000+ req/s with one core.

A cold run installing `express` takes ~12 seconds (fetching from upstream registry on-the-fly):

    npm install express  2.44s user 0.81s system 27% cpu 11.769 total

A warm run after all `express` assets are locally cached takes ~3 seconds:

    npm install express  2.43s user 0.78s system 115% cpu 2.768 total

## Implementation Caveats

You cannot use this mirror for publishing modules or user management, such
requests will be forwarded to the upstream registry for processing.


CHANGELOG.md
# Changelog

## 0.3.2
- Fixed a typo with stale resource config (#13)
- Make HTTPS the default upstream registry transport (#12)

## 0.3.1
- Fixed issue with config.json file not being found and issue where upstreamPort
  wasn't being honored ( #10 )
- Fixes an issue with Lactate cache expiry options leading to excessive CPU
  usage ( #11 )

## 0.3.0
- Large cleanup and refactor for new feature implementations.
- Now have the ability to serve stale resources from the cache while they cannot
  be refreshed from the upstream registry. (`--permit_stale_resources`)
- Support a list of blacklisted packages using semver ranges via a JSON
  configuration file. See an example configuration in the `examples/` folder.
  Enabled with `--package_blacklist <path to config file>`
- Added example server configuration in `examples/`
- Fix a bug where non 200 upstream registry responses were written to disk.
- Other minor bug fixes.

## 0.2.3
- Fix packages with multiple versions not being routed to the appropriate
  handler.

## 0.2.2
- Support version parsing when the package identifies itself with multiple
semvers.

## 0.2.1

- Ensure dots are parsed in the package name via HTTP routes.


test/npm.js
/**
 * Copyright (c) ActiveState 2013 - ALL RIGHTS RESERVED.
 */

'use strict';

var Config = require('../lib/config');
var Helpers = require('./helpers');
var Mkdirp = require('mkdirp');
var Npm = require('npm');

var npmLogLevel = 'warn';
var nodeModulesDir = '/tmp/npm-lazy-test-modules';

Npm.load({loglevel: npmLogLevel}, function (err, npm) {
    npm.config.set('registry', 'http://' + Config.serverAddress + ':' + Config.realExternalPort);
    npm.config.set('prefix', nodeModulesDir);
    npm.config.set('global', true);

    describe('It should make the test install folder', function () {
        it('should create ' + nodeModulesDir, function (done) {
            new Mkdirp(nodeModulesDir, done);
        });
    });

    describe('It should install packages using the npm module', function (){
        this.timeout(20000);

        it('should install npm', function (done) {
            npm.commands.install(['npm'], done);
        });

        var rand1 = Helpers.getRandomModule();
        it('should install a random module: ' + rand1, function (done) {
            this.timeout(1000 * 60 * 3);
            npm.commands.install([rand1], done);
        });

    });
});


test/runner.js
/**
 * Copyright (c) ActiveState 2013 - ALL RIGHTS RESERVED.
 */

'use strict';

var Helpers = require('./helpers');

var HOST = process.env.TEST_HOST || 'localhost';
var PORT = process.env.TEST_PORT || 2000;

var request = require('supertest');

request = request('http://' + HOST + ':' + PORT);

describe('GET /', function(){
    it('respond with json', function(done) {
        request.get('/')
        .set('Accept', 'application/json')
        .expect(200)
        .expect(/registry/, done);
    });
});

describe('GET /express', function(){
    it('responds 200', function(done) {
        request.get('/express')
        .set('Accept', 'application/json')
        .expect(200, done);
    });
});

describe('GET /express/latest', function(){
    it('respond 200 with Cache-Control header', function(done) {
        request.get('/express/latest')
        .set('Accept', 'application/json')
        .expect(200)
        .expect('Cache-Control', /max-age/, done);
    });
});

describe('GET /express/3.4.4', function(){
    it('gets the express 3.4.4 metadata', function(done) {
        request.get('/express/3.4.4')
        .expect(200)
        .expect('Cache-Control', /max-age/, done);
    });
});

describe('GET /multiparty/2.2.0', function(){
    it('responds 200 for the multiparty 2.2.0 metadata', function(done) {
        request.get('/multiparty/2.2.0')
        .expect(200)
        .expect('Cache-Control', /max-age/, done);
    });
});

describe('GET /multiparty/-/multiparty-2.2.0.tgz', function(){
    it('gets the multiparty 2.2.0 tarball', function(done) {

        this.timeout(10000);

        request.get('/multiparty/-/multiparty-2.2.0.tgz')
        .expect(200)
        .expect('Cache-Control', /max-age/)
        .expect('Content-Type', /application\/octet-stream/, done);
    });
});

describe('GET /multiparty/-/multiparty-2.222.0.tgz (fetch tarball)', function(){
    it('404 response on non-existing tarball', function(done) {
        request.get('/multiparty/-/multiparty-2.222.0.tgz')
        .expect(404, done);
    });
});

describe('GET //does/not/exist', function(){
    it('responds 404 on non-existing package', function(done) {
        request.get('//does/not/exist')
        .expect(404, done);
    });
});

describe('GET /doesnotexist', function(){
    it('responds 404 on non-existing package', function(done) {
        request.get('/doesnotexists')
        .expect(404, done);
    });
});
var randomModule = Helpers.getRandomModule();

describe('GET /' + randomModule, function(){
    it('returns a random module metadata', function(done) {
        request.get('/' + randomModule)
        .expect(200, done);
    });
});

describe('GET /raw-body/-/raw-body-1.1.2.tgz', function(){
    it('returns a tarball for package with multiple hyphens', function(done) {
        request.get('/raw-body/-/raw-body-1.1.2.tgz')
        .expect(200, done);
    });
});

describe('GET /dateformat/1.0.7-1.2.3', function(){
    it('returns metadata for a package with multiple versions', function(done) {
        request.get('/dateformat/1.0.7-1.2.3')
        .set('Accept', 'application/json')
        .expect('Content-Type', /json/)
        .expect(Helpers.tarballPointsToMirror)
        .expect(200, done);
    });
});



test/helpers.js
/**
 * Copyright (c) ActiveState 2013 - ALL RIGHTS RESERVED.
 */

'use strict';

var Config = require('../lib/config');
var Helpers = {};

Helpers.randomModules = [ 'express', 'supertest', 'mocha', 'hapi',
                          'coffee-script', 'lodash' ];

Helpers.getRandom = function(min, max) {
    return Math.floor(Math.random() * (max - min) + min);
};

Helpers.getRandomModule = function() {
    return Helpers.randomModules[Helpers.getRandom(0, Helpers.randomModules.length)];
};

Helpers.handleRes = function(err) {
    if (err) {
        throw err;
    }
};

Helpers.tarballPointsToMirror = function(res) {
    if (!res.body.dist.tarball) { return 'tarball field missing'; }
    if (res.body.dist.tarball.indexOf(Config.serverAddress + ':' + Config.realExternalPort) === -1) {
        return 'tarball URI does not map to the server external address:port';
    }
};

module.exports = Helpers;



server.js
/**
 * Copyright (c) ActiveState 2014 - ALL RIGHTS RESERVED.
 */

'use strict';

var Async = require('async');
var AsyncCache = require('async-cache');
var Cache = require('./lib/cache');
var Config = require('./lib/config');
var Constants = require('./lib/constants');
var Fs = require('fs');
var Handlers = require('./lib/handlers');
var Http = require('http');
var Https = require('https');
var Lactate = require('lactate');
var Log = require('log');
var Registry = require('./lib/registry');

/* Generic logger */
var log = new Log(Config.logLevel);

Config.log = log;

/* Upstream Proxy */
var registry = new Registry(Config);

/* HTTP asset caching server */
Config.cacheServer = Lactate.dir(Config.cacheDir, Config.cacheOptions);

/**
 * If the file cannot be found locally, proxy to the upstream registry as
 * a last resort
 */
Config.cacheServer.notFound(function (req, res) {
    registry.proxyUpstream(req, res);
});

/* Global async disk cache */
Config.cache.fsStats = new AsyncCache({
    max: 5000,
    maxAge: Config.cacheExpiry,
    load: function(path, cb) {

        log.info('Cache MISS: ', path);

        var cache = new Cache(Config);
        Async.waterfall([
            function (done) {
                cache.existsDisk(path, done);
            },
            function (exists, done) {
                if (exists) {
                    cache.diskExpired(path, done);
                } else {
                    done(null, null, null);
                }
            },
            function (expired, stats, done) {
                if (!expired && stats) {
                    done(null, Constants.CACHE_VALID);
                } else {
                    done(null, Constants.CACHE_NOT_EXIST);
                }
            }
        ],
        function (err, valid) {
            cb(err, valid);
        });
    }
});

/**
 * Primary HTTP/HTTPS request dispatcher
 *
 * @param {Object} req - HTTP.req
 * @param {Object} res - HTTP.res
 */
var serveRequest = function (req, res) {
    req.headers.host = Config.upstreamHost;

    log.info(req.method + ' ' + req.url);

    /* /package/latest */
    if (req.url.match(/^\/[_-\w.]+?\/latest\/?$/)) {
        Handlers.serveLatestPackageMeta(req, res, Config);

    /* /package/<semver> */
    } else if (req.url.match(/^\/[_-\w.]+?\/[0-9]+\.[0-9]+\.[0-9]+/)) {
        Handlers.servePackageVersionMeta(req, res, Config);

    /* /-/all/ */
    } else if (req.url.match(/\/-\/all\/.*/)) {
        registry.proxyUpstream(req, res);

    /* /-/<package name>-<version.tgz> */
    } else if (req.url.match(/^\/[_-\w.]+?\/-\/.*\.tgz/)) {
        Handlers.servePackageTarball(req, res, Config);

    /* /package */
    } else if (req.url.match(/^\/[_-\w.]+?$/)) {
        Handlers.servePackageMeta(req, res, Config);

    /* void */
    } else {
        registry.proxyUpstream(req, res);
    }
};

/**
 * The main HTTP server
 */
if (Config.httpEnabled) {
    var httpServer = Http.createServer(serveRequest);

    httpServer.listen(Config.httpPort, Config.bindAddress, function () {
        log.info('Lazy mirror (HTTP) is listening @ ' + Config.bindAddress + ':' + Config.httpPort + ' External host: ' + Config.serverAddress);
    });
}

/**
 * The main HTTPS server
 */
if (Config.httpsEnabled) {

    if (!Config.httpsKey || !Config.httpsCert) {
        throw new Error('Missing https_cert or http_key option');
    }

    var httpsOptions = {
        key: Fs.readFileSync(Config.httpsKey),
        cert: Fs.readFileSync(Config.httpsCert),
    };

    var httpsServer = Https.createServer(httpsOptions, serveRequest);

    httpsServer.listen(Config.httpsPort, Config.bindAddress, function () {
        log.info('Lazy mirror (HTTPS) is listening @ ' + Config.bindAddress + ':' + Config.httpsPort + ' External host: ' + Config.serverAddress);
    });
}




lib/cache.js
/**
 * Copyright (c) ActiveState 2013 - ALL RIGHTS RESERVED.
 */

'use strict';

var Async = require('async');
var Constants = require('./constants');
var Fs = require('fs');
var Mkdirp = require('mkdirp');
var Path = require('path');
var Url = require('url');

/**
 * Simple cache get/set operations for disk ops
 *
 * @constructor
 * @param {Object} opts - the main server config object
 * @returns {Cache}
 */
var Cache = function(opts) {

    if (!this instanceof Cache) {
        return new Cache(opts);
    }

    if (!opts) { return new Error('Options object not supplied'); }
    if (!opts.cacheDir) { return new Error('Cache directory not supplied'); }

    this.cacheDir = opts.cacheDir;
    this.config = opts;
    this.fsStats = opts.cache.fsStats;
    this.metaFolder = 'meta';
    this.tgzFolder = 'tgz';

    return this;
};

/**
 * Checks the mtime of the file has not exceeded the configured maximum expiry
 *
 * @param {string} path - file path
 * @callback cb - (err, bool, fs.Stats)
 */
Cache.prototype.diskExpired = function(path, cb) {

    var cache = this;

    Fs.stat(path, function(err, stats) {

        if (err) { return cb (err); }
        var now = Date.now();
        var mtime = stats.mtime.getTime();

        if ( (now - mtime) >= cache.config.cacheExpiry ) {
            return cb(null, true, stats);
        }

        return cb(null, false, stats);
    });
};

/**
 * Checks if the requested resource exists on the disk. This is an expensive I/O
 * op so should be checked with AsyncCache first.
 *
 * @param {string} path - file path
 * @callback cb - (err, bool)
 */
Cache.prototype.existsDisk = function(path, cb) {

    Fs.exists(path, function(exists) {
        cb(null, exists);
    });
};

/**
 * Checks if the requested resource is valid
 *
 * @param {Object} opts - type, name, version
 * @callback cb - (err, bool)
 */
Cache.prototype.validate = function(opts, cb) {

    if (!opts.type) { return cb(new Error('Type not supplied')); }
    if (!opts.name) { return cb(new Error('Name not supplied')); }
    if (!opts.version) { opts.version = ''; }

    var cache = this,
        path;
    if (opts.type === Constants.TYPE_META) {
        path = cache.getMetaPath(opts.name);
    } else if (opts.type === Constants.TYPE_TGZ) {
        path = cache.getTarballPath(opts.name, opts.version);
    } else {
        return cb(new Error('Unknown type: ' + opts.type));
    }
    cache.fsStats.get(path, cb);
};

/**
 * Overrides all the dist/tarball keys in the package metadata to point
 * to this mirror.
 *
 * @param {Object} pkg_json - The package metadata
 * @callback cb - (err, new_json)
 */
Cache.prototype.overrrideTarballUrls = function(pkgJson, cb) {

    var cache = this;

    // If the server is running with HTTPS enabled, default to that protocol
    var protocol = cache.config.httpsEnabled ? 'https:' : 'http';

    var changeDistUrl = function(ver, done) {
        ver = pkgJson.versions[ver];
        if (ver.dist && ver.dist.tarball) {
            var distUrl = Url.parse(ver.dist.tarball);
            distUrl.host = null;
            distUrl.hostname = cache.config.serverAddress;
            distUrl.port = cache.config.realExternalPort;
            distUrl.protocol = protocol;
            ver.dist.tarball = Url.format(distUrl);
            done();
        } else {
            done();
        }
    };

    if (!pkgJson.versions) { return cb(null, pkgJson); }

    Async.each(Object.keys(pkgJson.versions), changeDistUrl, function(err) {
        return cb(err, pkgJson);
    });

};

/**
 * Sets the metadata for a package
 *
 * @param {string} pkg - The package name
 * @param {Object} meta - The original package metadata
 * @callback cb - (err, pkg_json)
 */
Cache.prototype.setMeta = function(pkg, meta, cb) {

    var cache = this;
    var path = this.getMetaPath(pkg);
    var dir = Path.dirname(path);
    var pkgJson;

    Async.series([
        function(done) {
            new Mkdirp(dir, done);
        },
        function(done){
            Fs.writeFile(path + '.orig', JSON.stringify(meta), done);
        },
        function(done) {
            cache.overrrideTarballUrls(meta, function(err, json){
                if (err) { return done(err); }
                pkgJson = json;
                done();
            });
        },
        function(done) {
            Fs.writeFile(path, JSON.stringify(pkgJson), done);
        }
    ], function(err) {
        cb(err, pkgJson);
    });
};


/**
 * Sets the metadata for a specific package version
 *
 * @param {Object} pkg - The package (./lib/package) object
 * @param {Object} meta - The original package metadata
 * @callback cb - (err)
 */
Cache.prototype.setMetaVersion = function(pkg, meta, cb) {

    var path = this.getMetaVersionPath(pkg.name, pkg.version);
    var dir = Path.dirname(path);

    new Mkdirp(dir, function(err) {
        if (err) { return cb(err); }
        Fs.writeFile(path, JSON.stringify(meta), cb);
    });

};

/**
 * Gets the metadata filepath for a package
 *
 * @param {string} pkg - The package name
 * @returns {string}
 */
Cache.prototype.getMetaPath = function(name) {
    return Path.join(this.cacheDir, name, this.metaFolder, name + '.json');
};

/**
 * Gets the metadata filepath for a specific package version
 *
 * @param {string} pkg - The package name
 * @param {string} version - The package version
 * @returns {string}
 */
Cache.prototype.getMetaVersionPath = function(name, version) {
    return Path.join(this.cacheDir, name, this.metaFolder, name + '-' + version + '.json');
};

/**
 * Async wrapper for getMetaPath
 *
 * @param {string} pkg - The package name
 * @callback (err, path)
 */
Cache.prototype.getMeta = function(name, cb) {

    if (!name) { return cb(new Error('Name not supplied')); }
    return cb(null, this.getMetaPath(name));
};

/**
 * Async wrapper for getMetaVersionPath
 *
 * @param {string} pkg - The package name
 * @param {string} version - The package version
 * @callback (err, path)
 */
Cache.prototype.getMetaVersion = function(name, version, cb) {

    if (!name) { return cb(new Error('Name not supplied')); }
    if (!version) { return cb(new Error('Version not supplied')); }

    return cb(null, this.getMetaVersionPath(name, version));
};


/**
 * Gets the path to the tarball for a specific package version
 *
 * @param {string} pkg - The package name
 * @param {string} version - The package version
 * @returns {string}
 */
Cache.prototype.getTarballPath = function(name, version) {
    return Path.join(this.cacheDir, name, this.tgzFolder, name + '-' + version + '.tgz');
};

/**
 * Async wrapper for getTarballPath
 *
 * @param {string} pkg - The package name
 * @param {string} version - The package version
 * @callback (err, path)
 */
Cache.prototype.getTarball = function(name, version, cb) {

    if (!name) { return cb(new Error('Name not supplied')); }
    if (!version) { return cb(new Error('Version not supplied')); }

    return cb(null, this.getTarballPath(name, version));
};

module.exports = Cache;


lib/package.js
/**
 * Copyright (c) ActiveState 2014 - ALL RIGHTS RESERVED.
 */

'use strict';

var Async = require('async');
var Cache = require('./cache');
var Constants = require('./constants');
var Fs = require('fs');
var Registry = require('./registry');
var Semver = require('semver');

/**
 * Provides a package class for working with package metadata and binary blobs.
 *
 * @constructor
 * @param {Object} opts - Options
 * @returns {Package}
 */
var Package = function (opts) {

    if (!this instanceof Package) {
        return new Package(opts);
    }

    opts = opts || {};

    if (!opts.name) { return new Error('Package name not found'); }
    if (!opts.config && (!opts.config.upstreamHost || !opts.config.upstreamPort)) {
        return new Error('Registry is invalid: ' + opts.config);
    }

    this.name = opts.name;
    this.version = opts.version;

    this.config = opts.config;
    this.cache = new Cache(this.config);
    this.registry = new Registry(this.config);
    return this;

};

/**
 * Gets the entire metadata for a package and stores it on disk and returns
 * the path to the stored file (JSON).
 *
 * @callback done - {err, path}
 */
Package.prototype.getMeta = function(cb) {

    var pkg = this;

    if (pkg.version) {
        var blacklisted = pkg.validateAgainstBlacklist();
        if (blacklisted) { return cb(blacklisted); }
    }

    pkg.cache.validate({
            type: Constants.TYPE_META,
            name: pkg.name
        },
        function (err, validity) {
            if (err) { return cb(err); }

            if (validity === Constants.CACHE_EXPIRED && pkg.config.cachePermitStale) {
                Async.waterfall([
                    function (done) {
                        pkg.registry.getMeta(pkg.name, function (err, meta) {
                            done(null, meta);
                        });
                    },
                    function (meta, done) {
                        if (!meta) {
                            pkg.cache.getMeta(pkg.name, done);
                        } else {
                            done(null, meta);
                        }
                    }
                ], cb);
            } else if (validity === Constants.CACHE_VALID) {
                pkg.cache.getMeta(pkg.name, cb);
            } else {
                Async.waterfall([
                    function (done) {
                        pkg.registry.getMeta(pkg.name, done);
                    },
                    function (meta, done) {
                        pkg.cache.setMeta(pkg.name, meta, done);
                    },
                    function (meta, done) {
                        pkg.cache.getMeta(pkg.name, done);
                    }
                ], function (err, result) {
                    if (err) { return cb(err); }
                    pkg.config.cache.fsStats.set(result, true);
                    cb(null, result);
                });
            }
        });
};

/**
 * Extracts the latest version metadata from the master metadata file
 *
 * @callback done - {err, path}
 */
Package.prototype.getLatestVersionMeta = function(done) {

    var pkg = this;

    pkg.getMeta(function (err, metaPath) {
        if (err) { return done(err); }

        Fs.readFile(metaPath, function (err, meta){
            if (err) { return done(err); }
            meta = meta.toString();
            var data;

            try {
                data = JSON.parse(meta);
            } catch (e) {
                return done(e);
            }

            if (data['dist-tags'].latest) {
                pkg.version = data['dist-tags'].latest;
                var blacklisted = pkg.validateAgainstBlacklist();
                if (blacklisted) { return done(blacklisted); }
                pkg.getVersionMeta(done);
            } else {
                done(new Error('Cannot determine the latest version from the metadata'));
            }
        });
    });
};

/**
 * Extracts the version metadata from the master metadata file, and creates a
 * separate file if necessary, for serving with a cache server.
 *
 * @callback done - {err, path}
 */
Package.prototype.getVersionMeta = function(done) {

    var pkg = this;

    if (!pkg.version) { return done(new Error('This package has no version')); }

    var blacklisted = pkg.validateAgainstBlacklist();
    if (blacklisted) { return done(blacklisted); }

    pkg.getMeta(function (err, metaPath) {
        if (err) { return done(err); }

        Fs.readFile(metaPath, function (err, meta){
            if (err) { return done(err); }
            meta = meta.toString();
            var data;

            try {
                data = JSON.parse(meta);
            } catch (e) {
                return done(e);
            }

            if(data.versions[pkg.version]) {
                pkg.cache.setMetaVersion(pkg, data.versions[pkg.version], function (err){
                    if (err) { return done(err); }
                    pkg.cache.getMetaVersion(pkg.name, pkg.version, done);
                });
            } else {
                var error = new Error('Version not found');
                error.code = 404;
                done(error);
            }
        });
    });
};

/**
 * Assembles the package, checks the cache validity and reponds with the
 * path to the tarball
 *
 * @callback cb - {err, path}
 */
Package.prototype.getTarball = function(cb) {

    var pkg = this;

    var blacklisted = pkg.validateAgainstBlacklist();
    if (blacklisted) { return cb(blacklisted); }


    Async.waterfall([
        function(done) {
            pkg.cache.validate({
                type: Constants.TYPE_TGZ,
                name: pkg.name,
                version: pkg.version
            }, done);
        },
        function(validity, done){
            if (validity === Constants.CACHE_VALID || (validity === Constants.CACHE_EXPIRED && pkg.config.cachePermitStale)) {
                pkg.getTarballFromMeta(done);
            } else {
                pkg.registry.getTarball(pkg, done);
            }
        }
    ],
    cb);
};

/**
 * Extracts the tarball from the original JSON, and stores it in the package
 * directory, then returns the local path to it.
 * @callback cb = {err, path}
 */
Package.prototype.getTarballFromMeta = function (cb) {
    var pkg = this;

    pkg.getMeta(function(err, metaPath) {
        if (err) { return cb(err); }
        Fs.readFile(metaPath, function(err, meta){
            if (err) { return cb(err); }
            meta = meta.toString();
            var data;

            try {
                data = JSON.parse(meta);
            } catch (e) {
                return cb(e);
            }

            if (pkg.version) {
                if(data.versions[pkg.version]) {
                    pkg.cache.getTarball(pkg.name, pkg.version, cb);
                } else {
                    var error = new Error('Version not found');
                    error.code = 404;
                    return cb(error);
                }
            } else {
                return cb(new Error('Version not supplied'));
            }
        });
    });
};

/**
 * Validates the package name and semver against a supplied blacklist file.
 * Will return an error if the particular package is blacklisted, otherwise
 * undefined.
 *
 * @return {Error/undefined}
 */
Package.prototype.validateAgainstBlacklist = function () {

    var pkg = this;
    if (!pkg.version) {
        return new Error('Cannot validate a package with no version');
    }

    if (pkg.config.packageBlacklist && pkg.config.packageBlacklist[pkg.name]) {
        if (Semver.satisfies(pkg.version, pkg.config.packageBlacklist[pkg.name])) {
            return new Error('Package ' + pkg.name + ' version: ' + pkg.config.packageBlacklist[pkg.name] + ' is blacklisted');
        }
    }
    return;
};


module.exports = Package;


lib/config.js
/**
 * Copyright (c) ActiveState 2014 - ALL RIGHTS RESERVED.
 */

'use strict';

var Config = require('../config/config-defaults');
var PackageJSON = require('../package');
var Lactate = require('lactate');
var Npm = require('npm');
var FS = require('fs');

var Optimist = require('optimist')
.usage('Usage: $0 -p <port> -a <external_dns> -c <file cache dir>')
.describe({
    server_address: 'The external DNS name this mirror should serve on',
    bind_address: 'The local interface bind address',
    cache_dir: 'The full directory path the cache directory (no trailing slash)',
    http_port: 'The HTTP port for the lazy mirror to listen on',
    config: 'The path to the JSON configuration file',
    cache_expiry: 'Expire the upstream registry cache assets on this value (ms)',
    cache_mem: 'The maxmium size in MB of registry data to keep in memory. A larger allocation reduces disk hits and improves performance',
    http_enabled: 'Enable the HTTP server on -p, this is true if neither --http_enabled or --https_enabled are supplied',
    https_enabled: 'Enable the HTTPS server on --https_port',
    https_key: 'The path to the private SSL key',
    https_cert: 'The path to the private SSL certificate',
    https_port: 'The HTTPS port (requires --https_enabled)',
    http_proxy: 'Specify a HTTP proxy to traverse before making outbound requests',
    https_proxy: 'Specify a HTTPS proxy to traverse before making outbound requests',
    package_blacklist: 'Full path to a JSON configuration file of blacklisted packages',
    permit_stale_resources: 'Continue to serve stale resources if the upstream registry cannot be contacted',
    real_external_port: 'If this mirror is behind a proxy, set this flag to the real external port for client connections',
    upstream_host: 'The upstream registry host',
    upstream_port: 'The default upstream registry port',
    upstream_db_path: 'The default upstream registry database path, default: none (requires a leading slash)',
    upstream_use_https: 'Use HTTPS only when proxying to the upstream registry',
    upstream_verify_ssl: 'Verify the upstream registry\'s SSL certificates',
    log_level: 'Logging level',
    version: 'Print version and exit',
    help: 'This help screen'
})
.default({
    'server_address': 'localhost',
    'bind_address': '127.0.0.1',
    'cache_dir': '/tmp/files',
    'http_port': 2000,
    'cache_expiry': 24 * 60 * 60 * 1000, // 24 Hours
    'cache_mem': 200, // MB
    'https_port': 443,
    'upstream_host': 'registry.npmjs.org',
    'upstream_port': 80,
    'upstream_use_https': false,
    'upstream_verify_ssl': false,
    'log_level': 'info'
})
.alias({
    'server_address': 'a',
    'bind_address': 'b',
    'cache_dir': 'c',
    'http_port': ['port','p'],
    'config': 'C',
    'version': 'v',
    'help': 'h'
});

var Argv = Optimist.argv,
    config = {};

/* CLI specific */
if (Argv.help) {
    console.log(Optimist.help());
    process.exit();

} else if (Argv.version) {
    console.log('npm-lazy-mirror: ' + PackageJSON.version);
    process.exit();

} else if (Argv.config) {
    Config = JSON.parse(FS.readFileSync(Argv.config));
}

/* Local server options */
config.httpPort = Argv.http_port || Config.http_port;
config.serverAddress = Argv.server_address || Config.server_address;
config.bindAddress = Argv.bind_address || Config.bind_address;
config.httpEnabled = Argv.http_enabled || Config.http_enabled;

/* Local server HTTPS options */
config.httpsPort = Argv.https_port || Config.https_port;
config.httpsEnabled = Argv.https_enabled || Config.https_enabled;

if (!config.httpEnabled && !config.httpsEnabled) {
    config.httpEnabled = true;
}

config.realExternalPort = Argv.real_external_port || (config.httpsEnabled ? config.httpsPort : config.httpPort);

/* Logging */
config.logLevel = Argv.log_level || Config.log_level;

config.httpsKey = Argv.https_key || Config.https_key || null;
config.httpsCert = Argv.https_cert || Config.https_cert || null;

/* Upstream Registry */
config.upstreamHost = Argv.upstream_host || Config.upstream_host;
config.upstreamPort = Argv.upstream_port || Config.upstream_port;
config.upstreamUseHttps = Argv.upstream_use_https || Config.upstream_use_https;
config.upstreamVerifySSL = Argv.upstream_verify_ssl || Config.upstream_verify_ssl;
config.upstreamDBPath = Argv.upstream_db_path || Config.upstream_db_path || '';

/* Caching options */
config.cache = {};
config.cacheDir = Argv.cache_dir || Config.cache_dir;
config.cacheExpiry = Argv.cache_expiry || Config.cache_expiry
config.cacheMem = Argv.cache_mem || Config.cache_mem
config.cacheOptions = {
    cache: {
        expire: 60 * 60,
        max_keys: 500,
        max_size: config.cache_mem
    }
};
config.cachePermitStale = Argv.permit_stale_resources || Config.permit_stale_resources || false;

/* HTTP/S proxy */
config.getProxyConfig = function(cb) {
    Npm.load(function(err, npm) {
        if (err) { return (err); }
        return cb(null, {
            httpProxy: Argv.http_proxy || Config.http_proxy || Npm.config.get('proxy') || process.env.HTTP_PROXY,
            httpsProxy: Argv.https_proxy || Config.https_proxy || Npm.config.get('https-proxy') || process.env.HTTPS_PROXY
        });
    });
};

/* Package Blacklist */
config.packageBlacklist = (function() {
    if (Argv.package_blacklist || Config.package_blacklist) {
        return require(Argv.package_blacklist || Config.package_blacklist);
    }
})();

module.exports = config;
